### PR TITLE
Set up Storybook Documentation

### PR DESCRIPTION
**Pull request by hreineck**
Made inital setup for writing Storybook docs and
wrote up docs for components related to OSCAL Controls.

Added .stories.js files to the src/stories directory for
each component. These files render the components in the
Storybook stories.

Created a markdown page at src/stories/OSCALDocs.stories.mdx
that can serve as an initial framework for documenting the
Storybook page.
This will include a short description of each control
and the required and optional props for each one.

### CODE REVIEW

**Comment by hreineck** [comment]
To view the Storybook page, run:
`npm install
npm run storybook`

The .stories.js files in src/stories are responsible for rendering the docs.
They follow the format described in Storybook 's [documentation](https://storybook.js.org/docs/react/writing-stories/introduction).

**Comment by kylelaker** [comment]
It looks like maybe this is based off an old version of `develop` which is causing some of these pipeline failures. Could you try doing:

```
git pull --rebase origin develop
git push --force-with-lease origin EGRC-275
```

That might pull the commit that fixed these issues to this branch

**Comment by kylelaker** [comment]
hmm it looks like maybe that wasn't the issue... we've caused this issue to re-appear another way. interesting.

**Comment by tuckerzp** [comment]
Does running tests locally work @hreineck?

**Comment by hreineck** [comment]
> Does running tests locally work @hreineck?

No, i get this error that we had encountered before:
`Cannot find module '@testing-library/dom' from 'node_modules/@testing-library/user-event/dist/click.js'`

**Comment by tuckerzp** [comment]
@hreineck I get the same error. I have tried reinstalling `package-lock` & the `node-modules`. From what I have seen, the changes to `package.json` wouldn't cause this and I assume the `.story` files wouldn't either.

**Comment by rgauss** [comment]
Initial feedback after a quick look here:
- Overall, this is a great first start!!
- We should look to extend the existing granular test data to be able to support the import of it into stories so we have as little duplication as possible and are testing with data that then surfaces in the docs
- For components used across multiple consuming parent components, we should consider the OSCAL 'stack' from bottom up and describe the first story variant as something like `Default` or `ControlOnly` (rather than `No*`) then build with others that add modifications, constraints, etc. Simple cases at the top, increased complexity as we go down the variants (I keep using that term, blame it on the Loki series I guess)
- We should think about whether or not we want separate component stories for each component actually exposed in a file.
  - As an example, the `OSCALControlProse.js` file actually exposes two components `OSCALReplacedProseWithByComponentParameterValue` and `OSCALReplacedProseWithParameterLabel` because they share significant code, but are for different purposes.  Should those two components be separate stories?
 - Some of the examples use a story `title` with a prefix of `Components/...` instead of just the component which Storybook then renders under a Components group in the interface.  Perhaps we should investigate what's common and/or which method will lead to a better developer experience.
 - The Getting Started section should probably focus on how to build/access (npm registry) the library and how to use Storybook, rather than any repeat of the READMEs or components themselves

**Comment by hreineck** [comment]
> * The Getting Started section should probably focus on how to build/access (npm registry) the library and how to use Storybook, rather than any repeat of the READMEs or components themselves

So should we keep the markdown page listing the required/optional props for each component? Or is that
information covered by the stories themselves?

**Comment by tuckerzp** [comment]
I wonder if linting error is due to not committing the changes to `example/package-lock.json`?

**Comment by zclarkEDC** [comment]
> I wonder if linting error is due to not committing the changes to `example/package-lock.json`?

This is correct.
If you change the `package.json` , you need to commit the updated `package-lock.json` with it.

File: src/stories/OSCALControlImplementation.stories.js
Comment (Bronstrom) [inline src/stories/OSCALControlImplementation.stories.js:24]: Overall, this PR looks pretty good! Although, maybe we could avoid some code duplication in a couple parts that come directly from the test mock data. For example, this code here looks like it could be imported from `OSCALControlImplementationImpReq.test.js`.

File: package.json
Comment (hreineck) [reply package.json:60]: I think that got put there by the Storybook init script. Seems like it works fine just removing those lines.

File: src/stories/OSCALBackMatter.stories.js
Comment (rgauss) [inline src/stories/OSCALBackMatter.stories.js:11]: Can we change these names/implementations to:
* Default (includes title but no citation, description, or media type)
* With Description
* With Citation
* With Media Type
* Without Title

File: src/stories/OSCALBackMatter.stories.js
Comment (rgauss) [inline src/stories/OSCALBackMatter.stories.js:51]: Rather than repeating most of the data in each example, should the extended examples just add to the default?

We could do something like:
```javascript
const exampleBackMatterDescription = () => {
  const example = exampleBackMatter;
  example.resources[0].description = "some description";
  return example;
};
```
There may be a more elegant way to represent that.

File: src/stories/OSCALControl.stories.js
Comment (rgauss) [inline src/stories/OSCALControl.stories.js:5]: This should probably reflect the component name, so just `Control` rather than `Control Display`.

File: src/stories/OSCALControlImplementation.stories.js
Comment (rgauss) [reply src/stories/OSCALControlImplementation.stories.js:24]: Agreed, that what I was suggesting in my comment:
> We should look to extend the existing granular test data to be able to support the import of it into stories so we have as little duplication as possible and are testing with data that then surfaces in the docs.

File: src/stories/OSCALControl.stories.js
Comment (rgauss) [inline src/stories/OSCALControl.stories.js:112]: Can we have the default without constraints and add another of `WithParameterConstraints`?

File: src/stories/OSCALControl.stories.js
Comment (rgauss) [inline src/stories/OSCALControl.stories.js:13]: Should we call this `WithModifications`?

File: src/stories/OSCALControlPart.stories.js
Comment (rgauss) [inline src/stories/OSCALControlPart.stories.js:13]: Should this be `With ParameterConstraints`?

File: src/stories/OSCALControlPart.stories.js
Comment (rgauss) [inline src/stories/OSCALControlPart.stories.js:15]: Should this be `WithModifications`?

File: src/stories/OSCALControlProse.stories.js
Comment (rgauss) [inline src/stories/OSCALControlProse.stories.js:5]: I know I suggested we should consider separate stories for each component within a file, but I think in this case it's more confusing.

I'd combine into a single `Control Prose` story with no `Default` but:
* Parameters Replaced With Labels
* Parameters Replaced With By-Component Value
* With Parameter Constraints

File: src/stories/OSCALMetadata.stories.js
Comment (rgauss) [inline src/stories/OSCALMetadata.stories.js:15]: I think we can combine `MetadataParties` and `MetadataRoles` into a single `WithPartiesAndRoles`.

File: src/stories/OSCALModification.stories.js
Comment (rgauss) [inline src/stories/OSCALModification.stories.js:1]: I think this file needs to be renamed to `OSCALControlModification.stories.js`?

**Review by rgauss — COMMENTED** [review]
I think we're still waiting on consolidating with the test data.

We'll also want to update the root README with a link to the live Storybook deployment.

File: .github/workflows/gh-pages.yml
Comment (kylelaker) [inline .github/workflows/gh-pages.yml:7]: The *last* change before merging into develop needs to be to remove this line; I am going to request changes to make sure we keep a little red icon on the PR until this is done. But I also don't want to resolve this too soon because I want to see future changes to this branch still get built/published.

**Review by kylelaker — CHANGES_REQUESTED** [review]

File: src/stories/OSCALBackMatter.stories.js
Comment (hreineck) [reply src/stories/OSCALBackMatter.stories.js]: Something like this would affect all instances of `exampleBackMatter` because it's passed by reference. I'm not sure of an easy way to do deep copies here, but the back matter test data itself seems to avoid duplication pretty well already.

Comment (rgauss) [reply src/stories/OSCALBackMatter.stories.js]: Yes, that's why I mentioned just the "extended" examples, i.e. additive.  We could do deep copies for simple objects with JSON.parse and stringify though that can break down quickly with certain property types, or use a library like lodash, but I'm not sure either of those approaches would be worth it in this case.

I'm also fine with the current duplication, was just a thought.

**Comment by tuckerzp** [comment]
Is there a reason that linting and tests are not running?

File: src/stories/OSCALBackMatter.stories.js
Comment (rgauss) [inline src/stories/OSCALBackMatter.stories.js:21]: These variants are all within the Back Matter story so we can remove the `BackMatter` prefix from all of them.

File: src/stories/OSCALControlImplementation.stories.js
Comment (rgauss) [inline src/stories/OSCALControlImplementation.stories.js:23]: For consistency with the other components, lets rename this to `WithImplementedStatements`.

File: src/stories/OSCALControlPart.stories.js
Comment (rgauss) [reply src/stories/OSCALControlPart.stories.js:18]: Looks like this wasn't addressed?

File: src/stories/OSCALControlPart.stories.js
Comment (rgauss) [reply src/stories/OSCALControlPart.stories.js:20]: Looks like this wasn't addressed?

File: src/stories/OSCALControlProseByComponent.stories.js
Comment (rgauss) [inline src/stories/OSCALControlProseByComponent.stories.js:1]: I think we can delete this file now that we've consolidated in `OSCALControlProse.stories.js`?

File: src/stories/OSCALMetadata.stories.js
Comment (rgauss) [inline src/stories/OSCALMetadata.stories.js:17]: We don't need the `Metadata` prefix here, just `WithPartiesAndRoles`.

File: src/stories/OSCALSystemImplementation.stories.js
Comment (rgauss) [inline src/stories/OSCALSystemImplementation.stories.js:28]: I think these can all be consolidated into the single `Default`.

File: src/test-data/ModificationsData.js
Comment (rgauss) [inline src/test-data/ModificationsData.js:3]: Can we make these values something like `Some Prop Name` and `Some Prop Value` so it's a bit more obvious in the rending of the component?

File: README.md
Comment (rgauss) [inline README.md:26]: The README might flow better if we:
* Add the Storybook reference into what is currently Introduction
* Move what is currently Introduction into its own Components second level section before Setting Up

Something like:
# OSCAL React Library
A library of React components...
...
Take a tour ...

## Components
The library provides OSCAL UI developers...
Those components directly visualize...
**Documentation for the React components...**
The example folder...

## Setting Up
### Building and Running Locally
...

**Review by rgauss — CHANGES_REQUESTED** [review]

File: src/test-data/Urls.js
Comment (rgauss) [inline src/test-data/Urls.js:1]: Is it worth pulling these into their own file?

File: src/test-data/OtherData.js
Comment (rgauss) [inline src/test-data/OtherData.js:3]: This seems specific to modifications, should it be moved there?

File: src/test-data/OtherData.js
Comment (rgauss) [inline src/test-data/OtherData.js:1]: Should we rename this file from `OtherData` to something like `CommonData`?

**Review by rgauss — COMMENTED** [review]
A few more minor suggestions, but they don't have to hold up approval.

**Review by kylelaker — CHANGES_REQUESTED** [review]
I am still going to keep my "Request changes" until all other approvals are complete and we have made our last push to the branch; this is in regards to `on.push.branches[1]` and the need to remove that value but needing to _not_ impact the ability to see live deployments while developing.

File: src/test-data/Urls.js
Comment (hreineck) [reply src/test-data/Urls.js:1]: I think it's a bit cleaner to keep them in their own file, but if we'd rather pull it into something else that's easy enough to do.

File: src/test-data/OtherData.js
Comment (mikeisen1) [reply src/test-data/OtherData.js:3]: Are you referring to the adds const?

Comment (mikeisen1) [reply src/test-data/OtherData.js:3]: adds is only used in that file and isn't referenced anywhere else, that is why I placed it in OtherData.js (or whatever we rename that file to)

File: src/test-data/Urls.js
Comment (tuckerzp) [reply src/test-data/Urls.js:1]: Would it make more sense to have the urls that go with specific data go in that data file? The other urls could go with other common data.

For example: `backMatterTestUrl` goes into `BackMatterData.js `

File: src/test-data/OtherData.js
Comment (tuckerzp) [reply src/test-data/OtherData.js]: For #124 having the `modify` including the `adds` portion of a profile is used in places other than just `Profile` (`SSP` & `Component Definition`. Maybe having a `ModificationData` file makes sense? I am not sure if that is exactly what you mean @rgauss but it could prevent some duplication later.

Comment (hreineck) [reply src/test-data/OtherData.js]: It is related to modifications, but the only place it's used is the profile test. 
I think it'd technically be cleanest to modify the profile test to use the existing
test data in `test-data/ModificationsData` but that might be more trouble than it's worth.

File: src/test-data/Urls.js
Comment (zclarkEDC) [reply src/test-data/Urls.js:1]: Having the urls in their own file will be a nice fix if/when the urls change in the future.

**Review by zclarkEDC — CHANGES_REQUESTED** [review]
All local tests pass
```
Test Suites: 14 passed, 14 total
Tests:       66 passed, 66 total
Snapshots:   0 total
Time:        17.925 s
Ran all test suites.
```
Doing `npm run storybook` successfully builds all modules and opens the Storybook on localhost.

Console error when opening Control Prose -> Parameters Replaced With By-Component Value
console error
This issue is most likely caused in `OSCALControlProse.stories.js` line 51
```
ByComp.storyName = "Parameters Replaced With By-Component Value";
```

Removing this line caused the same error, however as you can see:
error 2
It changed the name to what looks like the correct naming scheme.

**Comment by tuckerzp** [comment]
>  We're trying to then pass that in as the `title` prop of `StyledTooltip` by assigning `description` to be equal to the returned object. This makes it undefined, and passing undefined as a prop gives the error.

When `statementByComponent` is de-structured, is `description` always undefined? 

> What should this title prop actually be? I thought of passing in something like the componentId prop as a workaround, but maybe that's not what's intended.

If `description` is not `undefined`, I would think we could still use it? Maybe if it is `undefined`, using `componentId` or some other variable would be good.  Maybe something like this?
```javascript
<StyledTooltip title={description ?? props.componentId}>
```

**Review by zclarkEDC — APPROVED** [review]

**Review by tuckerzp — COMMENTED** [review]
Everything looks good!

I was still curious about what everyone thought about Ray's recommendations of:
1. Putting urls into their relevant components
2. Creating a `modification` test data file and moving the `adds` & `modify` parameters to that file.

I was just hoping to get a final opinion on what to do for those conversations then I think we should be good to merge.

File: src/test-data/OtherData.js
Comment (hreineck) [reply src/test-data/OtherData.js]: I moved this to a more accurately named 'ProfileData.js` - hope that makes this more readable.

File: src/test-data/Urls.js
Comment (hreineck) [reply src/test-data/Urls.js:1]: I agree with @zclarkEDC - plus I think the Urls file is justified because some of the Urls are used across a couple different files.

**Review by rgauss — APPROVED** [review]

**Review by tuckerzp — APPROVED** [review]

**Review by kylelaker — APPROVED** [review]

**Pull request merged**
